features/370-auth-ui/design.md

```diff
@@ -0,0 +1,529 @@
+
+Auth UI as a OP has its own session management. Auth UI session is stateful and is stored in HTTP cookie. The session management API of Auth Gear cannot list, get or delete Auth UI session. Auth UI session and Auth Gear session are not interchangeable. This is analogous to a website utilizing Google as an OP cannot list the sessions the user have on `https://google.com`.
+
+### Diagram of New Relationship
+
+Below is a diagram to illustrate to the new relationship
+
+```
+                                        Skygear App
+
+
+                      Access Directly               Access either as First Party
+                                        +--------+      or with proper scopes
+                       +--------------->+  Data  +<----------------+
+                       |                +--------+                 |
+                       |                                           |
+                       |                                           |
+                 +-----+-----+       OP              RP     +------+------+
+                 |  Auth UI  +---------------+--------------+  Auth Gear  |
+                 +-----+-----+               |              +------+------+
+                       ^                     |                     ^
+                       |                     |                     |
+                       |                     |                     |
+                       |                     |                     |
+                       |                     v                     |
+             +---------+---------+   +-------+------+   +----------+----------+
+             |  Auth UI Session  |   |  API Client  |   |  Auth Gear Session  |
+             +---------+---------+   +-------+------+   +----------+----------+
+                       |                     ^                     |
+                       |                     |                     |
+Access via User Agent  |                     |  Identity with API  |
+                       |                     |                     |
+                       |                 +---+---+                 |
+                       +-----------------+  SDK  +-----------------+
+                                         +-------+
+```
+
+## Session compatibility with Auth Gear and the SDK
+
+In handling the Authentication Request in the Authorization Endpoint, Auth UI creates a session and store it in cookie. Finally it redirects the user agent back to the app with a authorization code.
+
+In a typical Authorization Code Flow, the app should send a Token Request to the Token Endpoint of Auth UI to obtain a Token Response. However, Token Response is incompatible with Auth Gear and the SDK, where the session transport can either be cookie or header.
```

Comment:
> Token Response is incompatible with Auth Gear and the SDK, where the session transport can either be cookie or header

Why? Token Response normally give a JWT token and after the JWT token is validated. it should be up to the RP to use whatever way to maintain the session transport? Or I didn't see something here?

Reply:
Ah I think I got what it meant after looking at the flow chart. Are we assuming the Auth UI is the OP? If yes the Token Endpoint should be at the Auth UI instead of Auth Gear?

Reply:
All OIDC endpoints implemented by Auth UI should be conforming to the spec. The Token endpoint is implemented by Auth UI.

The problem is the Token Response is JSON and the access token is inside the JSON body. If the SDK directly calls the Token endpoint of Auth UI, it is impossible to set access token in the cookie. The presence of the forwarding token endpoint in Auth Gear is solely for setting cookie and returning a backward compatible response the SDK expects. Of course, third party RP should call the Token endpoint of Auth UI directly.

Reply:
Possible to have the Auth Gear call token endpoint instead of client side SDK? seems more "normal" to the normal code flow.

Reply:
The flow is documented in [RFC7636 PKCE](https://tools.ietf.org/html/rfc7636#section-1.1). Only the client has the `code_verifier` so the following flow is impossible

```
 SDK                                                     Auth UI                                       Auth Gear
  +                                                         +                                                +
  |               Send Authentication Request               |                                                |
  +--------------------------------------------------------->                                                |
  |                                                         |                                                |
  |                           +-----------------------------+                                                |
  |                           |    Authenticate the user    |                                                |
  |                           +----------------------------->                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                      Write cookie                       |                                                |
  | Redirect to Auth Gear with Authentication Response      |                                                |
  |                                                         |                                                |
  |                                                         |             Follow HTTP redirect               |
  +--------------------------------------------------------------------------------------------------------->+
  |                                                         |                                                |
  |                                                         |              Stuck at the point                |
  |                                                         |  because Auth Gear does not have code verifier |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  |                                                         |                                                |
  +                                                         +                                                +
```